File: q76-100/q97.py
# 2020.01.04
# Problem Statement:
# https://leetcode.com/problems/interleaving-string/
# Referred to the solution here:
# https://leetcode.com/problems/interleaving-string/discuss/31879/My-DP-solution-in-C%2B%2B

import numpy as np

class Solution:
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        # check corner cases and do early returns
        if len(s1) == 0: return s2 == s3
        if len(s2) == 0: return s1 == s3
        if len(s1) + len(s2) != len(s3): return False

        # initialize dp structure
        # self.dp_answer[a][b] would represent if the first a elements in s1 
        # and the first b elements in s2 can form the first a+b elements in s3
        self.dp_answer = np.zeros((len(s1)+1, len(s2)+1))
        
        # start the loop
        for i in range(0, len(s1)+1):
            for j in range(0, len(s2)+1):
                # the first 0 element in s1 and first 0 element in s2 can definitely form the first 0 element in s3
                if i == 0 and j == 0:
                    self.dp_answer[i][j] = True
                elif i == 0:
                    self.dp_answer[i][j] = self.dp_answer[i][j-1] and s2[j-1] == s3[i+j-1]
                elif j == 0:
                    self.dp_answer[i][j] = self.dp_answer[i-1][j] and s1[i-1] == s3[i+j-1]
                else:
                    # the new element included can be either from s1 or s2
                    self.dp_answer[i][j] = (self.dp_answer[i][j-1] and s2[j-1] == s3[i+j-1]) or (self.dp_answer[i-1][j] and s1[i-1] == s3[i+j-1])       
        
        # return the last filled in in the dp structure
        return self.dp_answer[len(s1)][len(s2)]
        

# original version (personal solution), cannot pass some of the tests, time limit exceeded
class Solution:
    def isInterLeaveHelper(self, s1, s2, s3, take_s1):
        
        if len(s1) == 0: return s2 == s3
        if len(s2) == 0: return s1 == s3
        
        if take_s1:
            if s1[0] != s3[0]: return False
            else:
                i = 0
                while i < len(s1) and s1[i] == s3[i]:
                    i = i + 1
                
                for j in range(1, i+1):
                    self.answer = self.answer or self.isInterLeaveHelper(s1[j:], s2, s3[j:], False)
        else:
            if s2[0] != s3[0]: return False
            else:
                i = 0
                while i < len(s2) and s2[i] == s3[i]:
                    i = i +1
                
                for j in range(1, i+1):
                    self.answer = self.answer or self.isInterLeaveHelper(s1, s2[j:], s3[j:], True)
            
        return self.answer
    
    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        if len(s1) == 0: return s2 == s3
        if len(s2) == 0: return s1 == s3
        
        if len(s1) + len(s2) != len(s3): return False
        self.answer = False
        
        return self.isInterLeaveHelper(s1, s2, s3, True) or self.isInterLeaveHelper(s1, s2, s3, False)
        
        
        

File: Amazon/q97.java
// 2022.06.28
// Problem Statement:
// https://leetcode.com/problems/interleaving-string/

// idea: dfs, either s1 or s2 can be the first one in the interleaving,
// try both possibilities and if neither works, store false in the dp structure
class Solution {
    int [][] dp;
    public boolean isInterleaveHelper(String s1, String s2, String s3) {
        // base cases
        if (s1.length()==0) return s2.equals(s3);
        if (s2.length()==0) return s1.equals(s3);
        // check the dp structure (if substring s1 and substring s2 can form substring s3)
        if (dp[dp.length-s1.length()][dp[0].length-s2.length()]==1) return true;
        else if (dp[dp.length-s1.length()][dp[0].length-s2.length()]==2) return false;
        
        // try s1 as the head
        if (s1.charAt(0)==s3.charAt(0)) {
            boolean b1 = isInterleaveHelper(s1.substring(1), s2, s3.substring(1));
            if (b1) {
                dp[dp.length-s1.length()][dp[0].length-s2.length()] = 1;
                return true;
            }
        }

        // try s2 as the head
        if (s2.charAt(0)==s3.charAt(0)) {
            boolean b2 = isInterleaveHelper(s1, s2.substring(1), s3.substring(1));
            if (b2) {
                dp[dp.length-s1.length()][dp[0].length-s2.length()] = 1;
                return true;
            }
        }
        
        // does not work
        dp[dp.length-s1.length()][dp[0].length-s2.length()] = 2;
        return false;
    }
    
    public boolean isInterleave(String s1, String s2, String s3) {
        dp = new int [s1.length()][s2.length()]; // 0 uncalculated, 1 true, 2 false
        if (s1.length()==0) return s2.equals(s3);
        if (s2.length()==0) return s1.equals(s3);
        if (s3.length()!=(s1.length()+s2.length())) return false;
        return isInterleaveHelper(s1, s2, s3);
    }
}